### PR TITLE
transport: fix logical race in flow control

### PR DESCRIPTION
**Pull request by rhysh**
Remove the add and cancel methods of quotaPool. Their use is not required, and
leads to logical races when used concurrently from multiple goroutines. Rename
the reset method to add.

The typical way that a goroutine claims quota is to call the add method and
then to select on the channel returned by the acquire method. If two
goroutines are both trying to claim quota from a single quotaPool, the second
call to the add method can happen before the first attempt to read from the
channel. When that happens the second goroutine to attempt the channel read
will end up waiting for a very long time, in spite of its efforts to prepare
the channel for reading.

The quotaPool will always behave correctly when any positive quota is on the
channel rather than stored in the struct field. In the opposite case, when
positive quota is only in the struct field and not on the channel, users of
the quotaPool can fail to access available quota. Err on the side of storing
any positive quota in the channel.

This includes a reproducer for #632, which fails on many runs with this
package at v1.0.4. The frequency of the test failures depends on how stressed
the server is, since it's now effectively checking for weird interleavings of
goroutines. It passes reliably with these changes to the transport package.

The behavior described in #734 (an RPC with a streaming response hangs
unexpectedly) matches what I've seen in my programs, and what I see in the
test case added here. If it's a logical flow control bug, this change may well
fix it.

Updates #632
Updates #734

### CODE REVIEW

File: transport/http2_client.go [hunk 1]
```diff
@@ -614,9 +614,6 @@ func (t *http2Client) Write(s *Stream, data []byte, opts *Options) error {
 			// Wait until the transport has some quota to send the data.
 			tq, err := wait(s.ctx, s.done, s.goAway, t.shutdownChan, t.sendQuotaPool.acquire())
 			if err != nil {
-				if _, ok := err.(StreamError); ok || err == io.EOF {
-					t.sendQuotaPool.cancel()
-				}
```
Comment (iamqizhao) [inline transport/http2_client.go]: With this PR, I think we should be able to remove the "add" calls at line#607 and line#613. This applies to the server side too.

File: test/end2end_test.go [hunk 2]
```diff
@@ -3188,6 +3188,124 @@ func TestServerCredsDispatch(t *testing.T) {
 	}
 }
 
+func TestFlowControlIssue632(t *testing.T) {
```
Comment (iamqizhao) [inline test/end2end_test.go]: this is not a good name. how about "TestFlowControlRace"?

File: test/end2end_test.go [hunk 3]
```diff
@@ -3188,6 +3188,124 @@ func TestServerCredsDispatch(t *testing.T) {
...
+	testpb.RegisterTestServiceServer(s, &issue632server{
+		itemCount: itemCount,
+		itemSize:  itemSize,
+	})
+	defer s.Stop()
+
+	go s.Serve(lis)
+
+	ctx := context.Background()
+	ctx, _ = context.WithTimeout(ctx, totalTimeout)
```
Comment (iamqizhao) [inline test/end2end_test.go]: go test has timeout mechanism. You do not need to have your own here.

File: test/end2end_test.go [hunk 4]
```diff
@@ -3188,6 +3188,124 @@ func TestServerCredsDispatch(t *testing.T) {
...
+	defer func() {
+		if ctx.Err() == context.DeadlineExceeded {
+			t.Fatalf("test timed out")
+		}
+	}()
+
+	cc, err := grpc.Dial(lis.Addr().String(), grpc.WithInsecure())
```
Comment (iamqizhao) [inline test/end2end_test.go]: add grpc.WithBlock() to make the procedure more clear.

File: transport/http2_client.go (hunk 1 above)
Comment (rhysh) [inline transport/http2_client.go]: Done

File: test/end2end_test.go (hunk 3 above)
Comment (rhysh) [inline test/end2end_test.go]: Done

File: test/end2end_test.go (hunk 4 above)
Comment (rhysh) [inline test/end2end_test.go]: Done

File: test/end2end_test.go [hunk 5]
```diff
@@ -3188,6 +3188,117 @@ func TestServerCredsDispatch(t *testing.T) {
...
+			if failures >= maxFailures {
+				// Continue past the first failure to see if the connection is
+				// entirely broken, or if only a single RPC was affected
+				break
+			}
+		}
+	}
+}
+
+type issue632server struct {
```
Comment (iamqizhao) [inline test/end2end_test.go]: change the name too.

Comment (rhysh) [inline test/end2end_test.go]: Done

**Pull request merged**